### ISSUE TITLE
methodref.return.invalid

### ISSUE THREAD

**Issue by entonio** [typetools/checker-framework#1440]
This may be down to my ignorance of how generics work. But at first sight I think this code isn't wrong:

```java

import java.util.function.Function;

import org.checkerframework.checker.nullness.qual.Nullable;

public class Packet<T> {

    public static <T> Packet<T> get(T object) {
        return new Packet<>(object);
    }

    private final @Nullable T value;

    private Packet(@Nullable T value) {
        this.value = value;
    }

    public <U> @Nullable U map(Function<T, @Nullable U> mappingFunction) {
        if (value == null) {
            return (U) null;
        }
        return mappingFunction.apply(value);
    }

    public static void main(String[] args) {
        Object obj = 2.14;
        System.out.println("class: "
                + Packet.get(obj).map(Object::getClass)
        );
    }
}

```

Yet it produces the following:

```
Packet.java:28: warning: [methodref.return.invalid] Incompatible return type
                + Packet.get(obj).map(Object::getClass)
                                      ^
  Method
    @Initialized @NonNull Class<? extends @Initialized @Nullable Object> getClass(@UnknownInitialization @NonNull Object this) in java.lang.Object
  is not a valid method reference for
    @Initialized @Nullable Class<? extends @Initialized @NonNull Object> apply(@Initialized @NonNull Function<@Initialized @NonNull Object, @Initialized @Nullable Class<? extends @Initialized @NonNull Object>> this, @Initialized @NonNull Object p0) in java.util.function.Function
  found   : @Initialized @NonNull Class<? extends @Initialized @Nullable Object>
  required: @Initialized @Nullable Class<? extends @Initialized @NonNull Object>

```

What does this mean?

**Comment by entonio** [comment]
If I use `$ -> $.getClass()` in lieu of the method reference, it doesn't complain.

**Comment by wmdietl** [comment]
I just tried your example with the current 2.1.14 release and don't get an error.
Are you using the current version?
(Your signature for `map` is a bit strange, as both uses of `U` are annotated as `@Nullable`. I would remove the `@Nullable` in the parameter type.)

**Comment by smillst** [comment]
I was able to reproduce it with master.  This is a false positive.

**Comment by wmdietl** [comment]
Strange, cleaned and rebuilt everything and can now reproduce.

I don't think it's a false positive: https://github.com/typetools/checker-framework/blob/master/checker/jdk/nullness/src/java/lang/Object.java#L19
declares that the receiver for `Object::getClass` is `@UnknownInitialization @Raw` whereas `map` implicitly requires a fully initialized receiver.
This is not a problem in the lambda version, where the method is directly applied.

**Comment by smillst** [comment]
The error is about the return types, not the receiver types.
```
  found   : @Initialized @NonNull Class<? extends @Initialized @Nullable Object>
  required: @Initialized @Nullable Class<? extends @Initialized @NonNull Object>
```

The problem is that the type of `Object::getClass` should be inferred via Java 8 inference when inferring the type arguments of `map`.  Currently, type argument inference is using the wrong type for `Object::getClass`, so it inferred the wrong type argument for `U`.

**Comment by wmdietl** [comment]
`Object::getClass` doesn't have a method type parameter. So I'm not sure what type parameter you want to infer for that method.
The instantiation of `map`s `U` type parameter will be `Class`, but I don't see why that would restrict the type argument to `Class`.
So probably the special handling for the wildcard of `Object::getClass` is not applied to the method reference return type.

**Comment by smillst** [comment]
The type of a member reference depends on its assignment context, so the type of `Object::getClass` depends on the type of the parameter of `map` which depends on the inferred type argument for `U`. 

The current implementation of type argument inference uses the Java type for the tree `Object::getClass` with defaulted annotations (See TypeFromExpressionVisitor#visitMemberReference.) when inferring the type arguments for map.  That's why it is inferring the type argument for `U`  as  `Class<? extends @NonNull Object>` when it should be `Class<? extends @Nullable Object>`.

Bottom line is that this is a false positive that will be fixed when #979 is fixed.

**Comment by wmdietl** [comment]
I agree it's a false positive.
I don't think #979 alone will fix this. The change to the type argument to `Class` happens in `NullnessAnnotatedTypeFactory::adaptGetClassReturnTypeToReceiver` and we should make sure that method get's applied to method references.
It doesn't make sense to me to propose inferring `Class<? extends @Nullable Object>` when elsewhere we specifically change the bound to `@NonNull`.

**Comment by entonio** [comment]
> (Your signature for map is a bit strange, as both uses of U are annotated as @Nullable. I would remove the @Nullable in the parameter type.)

I'm saying two distinct things there:
1. `mappingFunction` returns U or null
2. `map` returns U or null

I can have 2 without 1, but if I use 1 without 2 then I would expect CF to complain. Am I thinking wrong? Either way, I'm specifying the method's interface for the benefit of client coders, not specifically for CF.

I've thought that putting an explicit annotation inside the method or class parametriser `<>` could be a way to say what the 'default' use of that parameter is within that method/class (so here I'd have `public <@Nullable U> U map(Function<T, U> mappingFunction)`), but even if that is/were so I've since found that it's a shorthand I'd rather not use.

I'm sorry about the rudimentary title. I looked for the warning and didn't find it, so I though I shouldn't try to guess where to be specific. I didn't think this was in any way related to `Object::getClass` itself.

**Comment by entonio** [comment]
In Swift I'd have written
```swift
public func map(mappingFunction: (T) -> U? ) -> U?
```

(I do understand `T?` is not exactly the same as `@Nullable T`)

**Comment by wmdietl** [comment]
@entonio Thanks for filing the issue! Sorry that @smillst and I dived into trying to fix it.
I agree that you could have 2 without 1 and I would probably have written that, just to give a real meaning to `U`. It's ok if you prefer this more explicit style.
